pkg/admission/no_downscale.go

```diff
@@ -0,0 +1,166 @@
+package admission
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	v1 "k8s.io/api/admission/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	autoscalingv1 "k8s.io/api/autoscaling/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	NoDownscaleLabelKey    = "grafana.com/no-downscale"
+	NoDownscaleLabelValue  = "true"
+	NoDownscaleWebhookPath = "/admission/no-downscale"
+)
+
+func NoDownscale(ctx context.Context, logger log.Logger, ar v1.AdmissionReview, api *kubernetes.Clientset) *v1.AdmissionResponse {
+	logger = log.With(logger, "name", ar.Request.Name, "resource", ar.Request.Resource.Resource)
+
+	oldObj, oldGVK, err := codecs.UniversalDeserializer().Decode(ar.Request.OldObject.Raw, nil, nil)
+	if err != nil {
+		return allowErr(logger, "can't decode old object, allowing the change", err)
+	}
+	logger = log.With(logger, "request_gvk", oldGVK)
+
+	oldReplicas, err := replicas(oldObj, oldGVK)
+	if err != nil {
+		return allowErr(logger, "can't get old replicas, allowing the change", err)
+	}
+	logger = log.With(logger, "old_replicas", int32PtrStr(oldReplicas))
+
+	newObj, newGVK, err := codecs.UniversalDeserializer().Decode(ar.Request.Object.Raw, nil, nil)
+	if err != nil {
+		return allowErr(logger, "can't decode new object, allowing the change", err)
+	}
+
+	newReplicas, err := replicas(newObj, newGVK)
+	if err != nil {
+		return allowErr(logger, "can't get new replicas, allowing the change", err)
+	}
+	logger = log.With(logger, "new_replicas", int32PtrStr(newReplicas))
+
+	// Changes from/to nil scale are not downscales strictly speaking.
+	if oldReplicas == nil {
+		return allowWarn(logger, "old replicas is nil, allowing the change")
+	}
+	if newReplicas == nil {
+		return allowWarn(logger, "new replicas is nil, allowing the change")
+	}
+
+	// If it's not a downscale, just log debug.
+	if *oldReplicas <= *newReplicas {
+		level.Debug(logger).Log("msg", "upscale allowed", "old_replicas", oldReplicas, "new_replicas", newReplicas)
```

Comment:
is this always an upscale?

Reply:
Right, improved feedback in https://github.com/grafana/rollout-operator/pull/25/commits/5481230bbb8e8eb66ce42e18429e5f9ff89b4299, also not warning if both old and new are nil.